File: cpp/quick_sort.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <chrono>
#include <filesystem>
using namespace std;
namespace fs = std::filesystem;

struct Record {
    int number;
    string text;
};

void quickSort(vector<Record>& arr, int low, int high) {
    if (low < high) {
        Record pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j].number <= pivot.number) {
                i++;
                swap(arr[i], arr[j]);
            }
        }
        swap(arr[i + 1], arr[high]);
        int pi = i + 1;

        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

vector<Record> readCSV(const string& filePath) {
    vector<Record> records;
    ifstream file(filePath);
    if (!file.is_open()) {
        cerr << "Failed to open file: " << filePath << endl;
        return records;
    }

    string line;
    while (getline(file, line)) {
        stringstream ss(line);
        string numStr, text;
        if (getline(ss, numStr, ',') && getline(ss, text)) {
            Record r;
            r.number = stoi(numStr);
            r.text = text;
            records.push_back(r);
        }
    }

    file.close();
    return records;
}

void writeCSV(const string& filePath, const vector<Record>& records) {
    ofstream out(filePath);
    for (const auto& r : records) {
        out << r.number << "," << r.text << "\n";
    }
    out.close();
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        cerr << "Usage: quick_sort <input_csv>" << endl;
        return 1;
    }

    string inputPath = argv[1];
    string fileName = fs::path(inputPath).filename().string();
    string outputPath = "cpp_output/quick_sort_" + fileName;

    // Read
    vector<Record> records = readCSV(inputPath);
    if (records.empty()) return 1;

    // Sort
    auto start = chrono::high_resolution_clock::now();
    quickSort(records, 0, records.size() - 1);
    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> elapsed = end - start;

    // Write
    fs::create_directory("cpp_output");
    writeCSV(outputPath, records);

    // Done
    cout << "Quick Sort completed in " << elapsed.count() << " seconds." << endl;
    cout << "Output written to " << outputPath << endl;
    return 0;
}


// .\cpp\quick_sort.exe data/dataset_60000000.csv

File: src/generator/sorting/quick_sort.py
import csv
import time
import os
import sys
from typing import List, Tuple

def read_csv(filepath: str) -> List[Tuple[int, str]]:
    data = []
    with open(filepath, 'r', encoding='utf-8') as f:
        reader = csv.reader(f)
        for row in reader:
            num = int(row[0])
            text = row[1]
            data.append((num, text))
    return data

def quick_sort(arr: List[Tuple[int, str]], low: int, high: int) -> None:
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

def partition(arr: List[Tuple[int, str]], low: int, high: int) -> int:
    pivot = arr[high][0]
    i = low - 1
    for j in range(low, high):
        if arr[j][0] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def write_csv(data: List[Tuple[int, str]], path: str) -> None:
    with open(path, 'w', encoding='utf-8', newline='') as f:
        writer = csv.writer(f)
        for row in data:
            writer.writerow(row)

def main():
    if len(sys.argv) != 3:
        print("Usage: python quick_sort.py <input_csv_path> <output_filename>")
        sys.exit(1)

    csv_path = sys.argv[1]
    output_name = sys.argv[2]

    os.makedirs("output", exist_ok=True)
    print(f"Reading data from {csv_path}...")
    data = read_csv(csv_path)

    print(f"Sorting {len(data):,} items using Quick Sort...")
    start_time = time.time()
    quick_sort(data, 0, len(data) - 1)
    end_time = time.time()

    duration = end_time - start_time
    print(f"✔ Quick Sort completed in {duration:.2f} seconds.")

    output_path = f"output/{output_name}"
    write_csv(data, output_path)
    print(f"✔ Output written to {output_path}")

if __name__ == "__main__":
    main()
